performance/multi-node/random_access.c

```diff
@@ -0,0 +1,1081 @@
+{
+	uint64_t *keys;
+	void **addresses;
+	uint64_t local_table_key, local_lock_key;
+	int i;
+	app_data_t *data = &fabric->data;
+	struct peer *peers = fabric->peers;
+
+	keys = calloc(num_ranks, sizeof(*keys));
+	assert(keys);
+
+	addresses = calloc(num_ranks, sizeof(*addresses));
+	assert(addresses);
+
+	local_table_key = fi_mr_key(data->l_table_mr);
+	local_lock_key = fi_mr_key(data->l_lock_mr);
+
+	DEBUG("Exchanging registration data\n");
+
+	ctpm_Barrier();
+
+	DEBUG("Allgather 1\n");
+	/* first, table addr */
+	ctpm_Allgather(&local_table_key, sizeof(local_table_key), keys);
+
+	for (i = 0; i < num_ranks; i++)
+		peers[i].data.r_table_key = keys[i];
+
+	ctpm_Barrier();
+
+	DEBUG("Allgather 2\n");
+	ctpm_Allgather(&local_lock_key, sizeof(local_lock_key), keys);
+
+	for (i = 0; i < num_ranks; i++)
+		peers[i].data.r_lock_key = keys[i];
+
+	ctpm_Barrier();
+
+	DEBUG("Allgather 3\n");
+
+	DEBUG("Sending table %p, size=%d, dest=%p\n", hpcc_table, sizeof(void *),
+			addresses);
+	ctpm_Allgather(&hpcc_table, sizeof(void *), addresses);
+	ctpm_Barrier();
+
+	for (i = 0; i < num_ranks; i++)
+		peers[i].data.r_table_addr = addresses[i];
+
+	ctpm_Barrier();
+
+	DEBUG("Allgather 4\n");DEBUG("Sending lock %p, size=%d, dest=%p\n",
+			hpcc_lock, sizeof(void *), addresses);
+	ctpm_Allgather(&hpcc_lock, sizeof(void *), addresses);
+
+	for (i = 0; i < num_ranks; i++)
+		peers[i].data.r_lock_addr = addresses[i];
+
+	ctpm_Barrier();
+	DEBUG("Finished exchanging data\n");
+}
+
+/* function declarations */
+CTPM_SIMPLE_TYPE_FUNC(int_sum_all, int, +);
+CTPM_SIMPLE_TYPE_FUNC(int64_sum_all, int64_t, +);
+
+static inline void cq_readerr(struct fid_cq *cq, const char *cq_str)
+{
+	struct fi_cq_err_entry cq_err;
+	const char *err_str;
+	int ret;
+
+	ret = fi_cq_readerr(cq, &cq_err, 0);
+	if (unlikely(ret < 0)) {
+		ct_print_fi_error("fi_cq_readerr", ret);
+		return;
+	}
+
+	err_str = fi_cq_strerror(cq, cq_err.prov_errno, cq_err.err_data, NULL, 0);
+	ERROR("%s: %d %s\n", cq_str, cq_err.err, fi_strerror(cq_err.err));
+	ERROR("%s: prov_err: %s (%d)\n", cq_str, err_str, cq_err.prov_errno);
+}
+
+/*
+ * fi_cq_err_entry can be cast to any CQ entry format.
+ */
+static inline int wait_for_comp(struct fid_cq *cq, int num_completions)
+{
+	struct fi_cq_err_entry comp;
+	int ret;
+
+	while (num_completions > 0) {
+		ret = fi_cq_read(cq, &comp, 1);
+		if (ret > 0) {
+			num_completions--;
+		} else if (unlikely(ret < 0 && ret != -FI_EAGAIN)) {
+			if (ret == -FI_EAVAIL) {
+				cq_readerr(cq, "cq");
+			} else {
+				ct_print_fi_error("fi_cq_read", ret);
+			}
+			return ret;
+		}
+	}
+	return 0;
+}
+
+static inline void lf_atomic_xor(void *addr, uint64_t val, int remote_rank)
+{
+	ssize_t rc;
+	struct peer *p = &prov.peers[remote_rank];
+	struct fid_ep *ep = p->ep;
+	struct fid_cq *scq = p->scq;
+	fi_addr_t dest_addr = prov.fi_addrs[remote_rank];
+	uint64_t rem_addr, key;
+	void *tmp = &scratch[0];
+	app_data_t *data = &prov.data;
+
+	// assume symmetric tables
```

Comment:
What do you mean by symmetric here?  Do you mean that they are the same size?  If so, that is more clear because symmetric is often used to refer to a *symmetric heap* in PGAS land.